### ISSUE TITLE
Upgrading to helm 3

### ISSUE THREAD

**Issue by minrk** [jupyterhub/mybinder.org-deploy#1542]
At some point, we are going to need to migrate our deployments to helm 3. Helm 3 has some changes about how it stores state and how it runs, so we'll need to be careful about doing upgrades to avoid losing stuff.

There's [some documentation](https://helm.sh/blog/migrate-from-helm-v2-to-helm-v3/) on how to do this successfully, and it might be better to do the migration manually, rather than on CI, once the update of CI is ready, but before it is merged.

**Comment by minrk** [comment]
refs from https://helm.sh/docs/topics/v2_v3_migration

- helm 2to3 plugin https://github.com/helm/helm-2to3
- https://helm.sh/blog/migrate-from-helm-v2-to-helm-v3

It's possible that we can run a manual 2to3 pass on each federation member and then "just" swap helm 2 for 3 and be off. Going to need some testing first, of course.

**Comment by choldgraf** [comment]
I believe that @consideRatio recently did some work in upgrading to helm 3 for JupyterHubs in general, perhaps he can help out? (apologies if I am mis-remembering that)

**Comment by sgibson91** [comment]
After upgrading Hub23 yesterday, I'm happy to give moving staging a go tomorrow

**Comment by sgibson91** [comment]
### Helm migration progress

| Federation member | 2to3 complete? | CI updated? |
| :--- | :---: | :---: |
| GKE staging | | |
| GKE prod | | |
| Turing | | |
| OVH | | |

*I assume GESIS will handle their deployment (if they haven't already done so)

**Comment by sgibson91** [comment]
In upgrading staging to helm3, we came across issues. `--force` in helm3 produces different behaviour to helm2 and immutable objects can no longer be updated, they have to be deleted and recreated. Hopefully this is a one-time thing that only needs to happen at the point of switching over/updating CI

<details><summary>Copy-pasting conversation from gitter</summary>

Sarah Gibson @sgibson91 Oct 26 14:20
Ok, seems the label selector is no having issues when trying to deploy to staging
https://github.com/jupyterhub/mybinder.org-deploy/pull/1670/checks?check_run_id=1309273607#step:14:60

Min RK @minrk 14:21
We used --force to deal with that before, did something change wrt force?

Sarah Gibson @sgibson91 Oct 26 14:21
Yes, @consideRatio advised not to use it with helm3

Min RK @minrk 14:22
But...we need it?
What's the alternative?
Looks like we need --recreate
Ah, looks like helm 3 has a separate flag to handle why we used force with helm 2
that's --recreate

Erik Sundell @consideRatio 14:24
@minrk Hmmmm, I don't find that in my helm3 as part of the helm upgrade --help list

Min RK @minrk 14:25
ok, I'm confused. Let's just put back --force, since it doesn't work without it
What has helm 3 changed about force that means we should stop using it?

Erik Sundell @consideRatio 14:25
  --force                        force resource updates through a replacement strategy
I'm not sure what happens when using --force with Helm3, will it replace a PVC that there is no change to, or, will it replace it if there is ANY change to it even though it was only a label change that was acceptable for example?

Min RK @minrk 14:26
I assume it's the same as helm 2, which means recreate when immutable fields are changed. Where did you see that it's more destructive than helm 2?

Erik Sundell @consideRatio 14:27
I'll find my way back to the issues, but this is my takeaway after reading lots of long threads about the topic.

I believe helm3 doesn't have --recreate


Min RK @minrk 14:27
ok, I must have misread an Issue

Erik Sundell @consideRatio 14:27
helm/helm#7811

Min RK @minrk 14:31
ok, then I don't understand what to do because we absolutely require the helm2 --force behavior and I don't understand if or how the helm 3 --force behavior is different.
I need to run and pickup at barnehage, bit will check in again tomorrow

Sarah Gibson @sgibson91 Oct 26 14:32
Np, thanks for your help today

Min RK @minrk 14:33
Thank you both!

Erik Sundell @consideRatio 14:34
Cya Min! Thank you for your help earlier btw!

helm3 --force is like kubectl apply --overwrite, while the previous helm2 --force was a bit different somehow. I'll look into how problematic --overwrite is or isn't

      --overwrite=true: Automatically resolve conflicts between the modified and live configuration by using values from the modified configuration
This sounds fine!
+1 for going with --force in this CI system again. Not 100% on --force in other situations, but feel less afraid reading that
I recall replacements like delete followed by create

Sarah Gibson @sgibson91 Oct 26 14:41
Thanks @consideRatio! I will add a commit changing this and see if we can finally get staging passing 

Erik Sundell @consideRatio 14:46


Sarah Gibson @sgibson91 Oct 26 14:47
Hmmm, still failing https://github.com/jupyterhub/mybinder.org-deploy/pull/1670/checks?check_run_id=1309410686#step:14:59

Erik Sundell @consideRatio 14:47
Thank you for your work on this @sgibson91
Oh wait what... hmm...
      --overwrite=true: Automatically resolve conflicts between the modified and live configuration by using values from
the modified configuration
Kubectl has overwrite to true by default, so ... helm upgrade --force is kube kubectl apply by default?
https://banzaicloud.com/blog/helm3-the-good-the-bad-and-the-ugly/#changing-the-unchangeable

Sarah Gibson @sgibson91 Oct 26 14:53
Soooooo, it's impossible?

Erik Sundell @consideRatio 14:53
"If you want to modify the role binding, my suggestion would be to delete it from the cluster first before upgrading."

Sarah Gibson @sgibson91 Oct 26 14:56
Yes, but judging by the error message, I'd have to delete everything :/

Erik Sundell @consideRatio 15:01
So i've kept reading about this, for like the 3-5th time, and always left feeling not confident.

It is a complicated issue, and they are not sure how to make helm3 function after some issues in helm2. (helm/helm#7082).

I think the closest option we have is: helm delete --keep-history && helm install --replace to always work, as suggested in that PR

in that issue i mean
To conclude, Helm2 had a functionality that Helm3 doesn't - the ability to try to carefully update a resource, but when that fail, be able to delete it and create it again. Helm 3 doesn't have this, and will simply fail when it is required to delete->create a resource to update it without any ability to workaround it with helm upgrade. The fallback is to manually delete the resources and then try an upgrade again.

Sarah Gibson @sgibson91 Oct 26 15:05
Hmmmm, ok
I think I'm going to meditate on this and come back to it tomorrow

Erik Sundell @consideRatio 15:09
Error: UPGRADE FAILED:
failed to replace object: PersistentVolumeClaim "hub-db-dir" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims
failed to replace object: PersistentVolumeClaim "staging-grafana" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims
failed to replace object: PersistentVolumeClaim "staging-prometheus-server" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims
failed to replace object: Service "hub" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "proxy-api" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "proxy-public" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "binder" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "staging-grafana" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "staging-ingress-nginx-controller-metrics" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "staging-ingress-nginx-controller" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "staging-ingress-nginx-defaultbackend" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "staging-kube-state-metrics" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "staging-prometheus-server" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "federation-redirect" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "gcs-proxy" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "matomo" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "proxy-patches" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Service "redirector" is invalid: spec.clusterIP: Invalid value: "": field is immutable
failed to replace object: Deployment.apps "analytics-publisher" is invalid: spec.selector: Invalid value: v1.LabelSelector***MatchLabels:map[string]string***"app":"analytics-publisher", "heritage":"Helm", "release":"staging"***, MatchExpressions:[]v1.LabelSelectorRequirement(nil)***: field is immutable
failed to replace object: Deployment.apps "federation-redirect" is invalid: spec.selector: Invalid value: v1.LabelSelector***MatchLabels:map[string]string***"app":"federation-redirect", "heritage":"Helm", "release":"staging"***, MatchExpressions:[]v1.LabelSelectorRequirement(nil)***: field is immutable
failed to replace object: Deployment.apps "gcs-proxy" is invalid: spec.selector: Invalid value: v1.LabelSelector***MatchLabels:map[string]string***"app":"gcs-proxy", "component":"nginx", "heritage":"Helm", "release":"staging"***, MatchExpressions:[]v1.LabelSelectorRequirement(nil)***: field is immutable
failed to replace object: Deployment.apps "matomo" is invalid: spec.selector: Invalid value: v1.LabelSelector***MatchLabels:map[string]string***"app":"matomo", "component":"matomo", "heritage":"Helm", "release":"staging"***, MatchExpressions:[]v1.LabelSelectorRequirement(nil)***: field is immutable
failed to replace object: Deployment.apps "matomo-mysqld-exporter" is invalid: spec.selector: Invalid value: v1.LabelSelector***MatchLabels:map[string]string***"app":"matomo", "component":"mysqld-exporter", "heritage":"Helm", "release":"staging"***, MatchExpressions:[]v1.LabelSelectorRequirement(nil)***: field is immutable
I split the error message apart using && as an indicator of separate errors.

So, I think if you do ...

kubectl delete pvc hub-db-dir staging-grafana staging-prometheus-server
kubectl delete svc hub proxy-api proxy-public binder staging-grafana staging-ingress-nginx-controller-metrics staging-ingress-nginx-controller staging-ingress-nginx-defaultbackend staging-kube-state-metrics staging-prometheus-server federation-redirect gcs-proxy matomo proxy-patches redirector
kubectl delete deploy analytics-publisher federation-redirect gcs-proxy matomo matomo-mysqld-exporter
Then you are ready for a helm upgrade again
:'(

Sarah Gibson @sgibson91 Oct 26 15:12
Will that have to go in the CI every time though, or just for this migration? And what information do we lose (if any) by deleting those e.g. pvcs?

Erik Sundell @consideRatio 15:14
This needs to be done whenever the Helm chart upgrade helm upgrade does leads to a change of something that k8s sais requires a recreation of the resource
If you for example do a change to a k8s deployment resource matchLabels, then you require a recreation

Sarah Gibson @sgibson91 Oct 26 15:15
So could we not put it in the CI and instead have a section in the site reliability guide saying this needs to be done?

Erik Sundell @consideRatio 15:18
Hmmmm I mean, if the CI is updating something that requires delete->create, and that was a PVC, then you ended up deleting the PVC which may be very problematic as the storage is whiped.

So, okay to always recreate? Perhaps, not sure what resources we have that could do bad when recreated. But, for a real deployment when we assume we have valuable things in requested storage disks, then doing for example helm delete && helm install in sequence is madness


Sarah Gibson @sgibson91 Oct 26 15:28
Yeah, I'm just wondering if we do actually require delete->create every time?

Erik Sundell @consideRatio 15:33
Then something is wrong
If z2jh upgrade make us require that, i make a breaking change note

Sarah Gibson @sgibson91 Oct 26 15:34
Sorry, I'm confused now, what's wrong?
Here's what I'm thinking:
1) Maybe (I don't know this for certain) not every helm upgrade we do on mybinder.org actually involves the need to delete and reinstall, so we don't put delete and reinstall in the CI
2) For the few cases that do require a delete and reinstall, we have docs in the SRE telling people how to do that, i.e. it becomes a manual step that happens infrequently

Sarah Gibson @sgibson91 Oct 26 15:39
Or I'm wrong, a delete and reinstall is required with every helm upgrade and we have to have some ridiculous CI build?

Erik Sundell @consideRatio 15:43
Assume you only had z2jh, then, helm upgrade would not need --force etc, because it doesnt change resources that are immutable
So, if the CI system running helm upgrade thinks it required, then something is wrong
 1 reply
And, i saw hub resources in that list. So, what is going on?
I think it should be like step 1/2 described abobe by you, but what cause helm to think wr need tl recreate resources because it think they are changing to much? Is something modofying the resources mid way?

Sarah Gibson @sgibson91 Oct 26 15:49
This is the first upgrade since migrating to helm3 though, so I wonder if that's the reason this step is required now? I had a similar issue with Hub23 but assumed I had done something wrong

Min RK @minrk 08:02
losing PVCs on mybinder.org is dramatically less disastrous than for most JupyterHub deployments. Primarily because our only PVCs are Hub state (would lose all current sessions, which is a major but short-lived disruption) and grafana/prometheus. Our grafana charts are backed up (not constantly, but recently) and prometheus is ephemeral so not super valuable. It has happened at least once on a prometheus upgrade and nobody really noticed.

Min RK @minrk 08:17
We appear to be having these conflicts due to the change in the heritage label from Tiller to Helm, causing a one-time conflict on all resources

Erik Sundell @consideRatio 08:22
Oh heritage is not suppored to be a matchLabel btw

Min RK @minrk 08:23
it's not a matchLabel, is it?
it super bugs me that helm upgrade --dry-run does not include conflict checks

Min RK @minrk 08:31
Ah, I see we have heritage in matchLabel. I don't understand what it means to be "not supported" as a matchLabel
I'm not sure if that's a source of conflict? That wouldn't make sense

Min RK @minrk 08:38
ok, it does appear that matchLabels is the source of some of these conflicts. Removing them may help

Sarah Gibson @sgibson91 08:42
Would that be something like
kubectl delete matchLabel label1 label2... -n staging

Min RK @minrk 08:43
I think it's delete & recreate the deployment objects
It should be a one-time thing, and the --force in our helm2 deployments should mean that they won't have the same conflict
Is it okay if I try a --force upgrade on staging to see if we lose the grafana PV? There's nothing in the staging one worth keeping
I'm curious if it will survive or not

Sarah Gibson @sgibson91 08:44
I already pushed a commit that added --force yesterday and it didn't work

Min RK @minrk 08:45
ah, ok
</details>

**Comment by sgibson91** [comment]
**TL;DR:** helm 2to3 may not actually be totally compatible with most deployed things and the best route forward may be to delete the helm2 deployment completely and redeploy with helm3. This is ok for staging and Turing/OVH because they're either not a live service or can be removed from the federation while work is underway. A different strategy will need to be implemented for prod so I will leave this cluster til last.

**Note:** This "delete helm2 and redeploy with helm3" strategy is ultimately what I ended up doing for Hub23 but I thought it was because I made a mistake 😬 

<details><summary>Some more discussion from gitter</summary>

Min RK @minrk 08:47
Ah, I see. --force results in different conflicts. Fun!

Sarah Gibson @sgibson91 08:51
I feel like if I run the below, a lot of these will recreate themselves automatically, i.e. before I can run helm upgrade. Do we think that will be problematic?
kubectl delete pvc hub-db-dir staging-grafana staging-prometheus-server
kubectl delete svc hub proxy-api proxy-public binder staging-grafana staging-ingress-nginx-controller-metrics staging-ingress-nginx-controller staging-ingress-nginx-defaultbackend staging-kube-state-metrics staging-prometheus-server federation-redirect gcs-proxy matomo proxy-patches redirector
kubectl delete deploy analytics-publisher federation-redirect gcs-proxy matomo matomo-mysqld-exporter

Min RK @minrk 08:51
It'll bring down the whole service
Seems like the service conflict will happen every time on --force

Sarah Gibson @sgibson91 08:52
Yes, aside from that ;)

Min RK @minrk 08:52
IIUC
whereas the non-force conflicts will happen just once

Sarah Gibson @sgibson91 08:52
Ok, so revert my last commit on the PR

Min RK @minrk 08:53
Maybe? I'm still trying to understand

Erik Sundell @consideRatio 08:53
@minrk i meant heritage isnt supposed to be a matchLabel according to best practices of Helm charts, so in z2jh we have avoided that, but ive seen it show up here and there in other charts

Min RK @minrk 08:53
--force is showing a conflict on unchanged resources (clusterIP) which worries me more than the conflicts on matchLabels, which we can fix
The remaining problem is that the grafana PVC uses heritage in a Label, which I guess is a bad practice, despite being a recommended best practice
~all of this would have been avoided if helm3 didn't decide to change the heritage value

Erik Sundell @consideRatio 08:55
To have a label is one thing, but to use heritage as a matchLabel / label selector, isnt

Min RK @minrk 08:55
Yes, but it's the label causing the conflict on the PVC
Most of the conflicts are on matchLabel, which we can fix. One remaining is on the grafana PVC, which is just a label
This one:
 cannot patch "staging-grafana" with kind PersistentVolumeClaim: PersistentVolumeClaim "staging-grafana" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims

Erik Sundell @consideRatio 08:56
Oh

Min RK @minrk 08:57
and the whole diff:
staging, staging-grafana, PersistentVolumeClaim (v1) has changed:
  # Source: mybinder/charts/grafana/templates/pvc.yaml
  apiVersion: v1
  kind: PersistentVolumeClaim
  metadata:
    name: staging-grafana
    namespace: staging
    labels:
      app: grafana
      chart: grafana-3.10.1
      release: staging
-     heritage: Tiller
+     heritage: Helm
    finalizers:
      - kubernetes.io/pvc-protection
  spec:
    accessModes:
      - "ReadWriteOnce"
    resources:
      requests:
        storage: "1Gi"
    storageClassName:

Erik Sundell @consideRatio 08:57
So: "kubectl label pvc" is forbidden?

Min RK @minrk 08:58
while it is bound, I guess

Erik Sundell @consideRatio 08:58
Wow, okay

Min RK @minrk 09:00
Here's what I don't understand: the prometheus-server PVC has heritage: Helm label already, but the grafana PVC has heritage: Tiller
I can't find anything different about them
maybe it was a coincidence that the prometheus server was not running when it patched the PVC? I really don't understand this.
Seems like the helm 2 to 3 upgrade process, while it's smooth to start, is a total lie that it's actually compatible with deployments deployed with helm 2, and everyone should just delete and deploy fresh with helm 3.
And this is ~100% due to the change in the heritage label.

Sarah Gibson @sgibson91 09:03
That's ultimately what I ended up doing with Hub23 
But surely we can't do that with prod?!

Min RK @minrk 09:03
Yeah, I think it might be the right thing to do
I'm not sure. I have other, more time sensitive stuff today, but I'll check back in this afternoon

Sarah Gibson @sgibson91 09:04
I'm feeling I'll leave prod to be the last cluster to be updated

Min RK @minrk 09:04
I think that's a good plan

Sarah Gibson @sgibson91 09:04
At least with Turing/OVH, we can remove them from the federation
</details>

**Comment by sgibson91** [comment]
[This commit](https://github.com/jupyterhub/mybinder.org-deploy/pull/1670/commits/f5d0b3640d1a878bc07e45e0bc0169069136ea09) solved the issue in the end. Helm3 doesn't like empty variables.

Plan going forward:

1. Merge #1670 and babysit
2. Let things settle for a couple of days
3. Work on Turing next. See if any other issues turn up during that upgrade.
   - If they do, a potential fix may be to upgrade our prometheus and grafana dependencies before moving onto prod as most of the issues came from the analytics publisher, "stuff that just runs on staging/prod"

**Comment by sgibson91** [comment]
Turing upgrade now complete! 2 down, 2 to go

**Comment by sgibson91** [comment]
As of yesterday, OVH is now running Helm 3! :tada:

If anyone has some recommendations on how to smoothly do the upgrade on prod, I'd love to hear! For the other clusters, I just removed them from the federation before I started tinkering. But that's not an option with prod.

**Comment by sgibson91** [comment]
> If they do, a potential fix may be to upgrade our prometheus and grafana dependencies before moving onto prod as most of the issues came from the analytics publisher, "stuff that just runs on staging/prod"

I actually think this would be a good idea as this is what @consideRatio and I ended up debugging yesterday (in the first place, at least 😂 )

**Comment by consideRatio** [comment]
- [x] grafana and prometheus upgraded a bit but not all the way as that requires Helm3 now, and I've made notes in requirements.yaml on future upgrades with Helm 3.
- [x] GKE prod upgraded to use helm3 manually with 2to3 and a trial deployment
- [x] CI system upgraded to use helm3 for GKE prod also
- [x] Helm2 revision data migrated to Helm3 deleted in:
  - [x] GKE Prod
  - [ ] OVH
  - [ ] Staging
  - [ ] Turing
- [x] Helm2 tiller removed in:
  - [x] GKE Prod
  - [ ] OVH
  - [ ] Staging
  - [ ] Turing

**Comment by betatim** [comment]
Nice work!

I think Gesis has been using helm 3 for quite a while already. It isn't deployed from this repository but I don't know what exactly they use. A question for @bitnik and @MridulS

**Comment by sgibson91** [comment]
> * [x]  Manually cleanup Helm2 revision data already migrated to Helm3:
>   * [ ]  Turing (Ping @sgibson91)
> * [x]  Manually cleanup Helm2 tiller deployment and ServiceAccount:
>   * [ ]  Turing (Ping @sgibson91)

Are these steps something different to running `helm3 2to3 cleanup`? If not, they were done back when I did the migration.

**Comment by consideRatio** [comment]
@sgibson91 i needed to manually delete the tiller SA, but tiller deploy was deleted as part of the helm 2to3 cleanup

**Comment by sgibson91** [comment]
OK, well I plan on completely destroying and redeploying the cluster tomorrow to try and fix the DNS issue - so there definitely shouldn't be any helm2 stuff after that!

**Comment by sgibson91** [comment]
Tiller service account deleted on Turing cluster

**Comment by MridulS** [comment]
> Is Gesis not using the mybinder.org-deploy Helm chart?

Yeah, we don't use the mybinder.org-deploy Helm chart. We already are at helm3 for some time now :)

**Comment by sgibson91** [comment]
Reopened as I realised #1740 is still open
